sql/src/main/java/io/crate/planner/node/ddl/CreateSnapshotPlan.java

```diff
@@ -0,0 +1,207 @@
+import io.crate.expression.symbol.Symbol;
+import io.crate.metadata.CoordinatorTxnCtx;
+import io.crate.metadata.Functions;
+import io.crate.metadata.Schemas;
+import io.crate.metadata.doc.DocTableInfo;
+import io.crate.metadata.table.Operation;
+import io.crate.metadata.table.SchemaInfo;
+import io.crate.metadata.table.TableInfo;
+import io.crate.planner.DependencyCarrier;
+import io.crate.planner.Plan;
+import io.crate.planner.PlannerContext;
+import io.crate.planner.operators.SubQueryResults;
+import io.crate.sql.tree.Table;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.admin.cluster.snapshots.create.CreateSnapshotRequest;
+import org.elasticsearch.action.support.IndicesOptions;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.snapshots.SnapshotInfo;
+import org.elasticsearch.snapshots.SnapshotState;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.function.Function;
+
+import static io.crate.analyze.PartitionPropertiesAnalyzer.toPartitionName;
+import static io.crate.analyze.SnapshotSettings.IGNORE_UNAVAILABLE;
+import static io.crate.analyze.SnapshotSettings.WAIT_FOR_COMPLETION;
+
+public class CreateSnapshotPlan implements Plan {
+
+    private static final Logger LOGGER = LogManager.getLogger(CreateSnapshotPlan.class);
+
+
+    private final AnalyzedCreateSnapshot createSnapshot;
+
+    public CreateSnapshotPlan(AnalyzedCreateSnapshot createSnapshot) {
+        this.createSnapshot = createSnapshot;
+    }
+
+    @Override
+    public StatementType type() {
+        return StatementType.DDL;
+    }
+
+    @Override
+    public void executeOrFail(DependencyCarrier dependencies,
+                              PlannerContext plannerContext,
+                              RowConsumer consumer,
+                              Row parameters,
+                              SubQueryResults subQueryResults) {
+        CreateSnapshotRequest request = createRequest(
+            createSnapshot,
+            plannerContext.transactionContext(),
+            plannerContext.functions(),
+            parameters,
+            subQueryResults,
+            dependencies.schemas());
+
+        var transportCreateSnapshotAction = dependencies.transportActionProvider().transportCreateSnapshotAction();
+        transportCreateSnapshotAction.execute(
+            request,
+            new OneRowActionListener<>(
+                consumer,
+                response -> {
+                    SnapshotInfo snapshotInfo = response.getSnapshotInfo();
+                    if (snapshotInfo != null &&  // if wait_for_completion is false, the snapshotInfo is null
+                        snapshotInfo.state() == SnapshotState.FAILED) {
+                        // fail request if snapshot creation failed
+                        String reason = response.getSnapshotInfo().reason()
+                            .replaceAll("Index", "Table")
```

Comment:
Maybe we could follow up on this later and change the reason definitions within SnapshotInfo to make this unnecessary

Reply:
sure, will follow up on in a separate PR.